File: packages/expo-image-manipulator/android/src/main/java/expo/modules/imagemanipulator/ImageManipulatorModule.java
package expo.modules.imagemanipulator;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.Base64;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;

import org.unimodules.core.ExportedModule;
import org.unimodules.core.ModuleRegistry;
import org.unimodules.core.Promise;
import org.unimodules.core.arguments.ReadableArguments;
import org.unimodules.core.interfaces.ExpoMethod;
import org.unimodules.interfaces.imageloader.ImageLoader;

import expo.modules.imagemanipulator.arguments.Action;
import expo.modules.imagemanipulator.arguments.ActionCrop;
import expo.modules.imagemanipulator.arguments.ActionFlip;
import expo.modules.imagemanipulator.arguments.ActionResize;
import expo.modules.imagemanipulator.arguments.SaveOptions;

public class ImageManipulatorModule extends ExportedModule {
  private static final String TAG = "ExpoImageManipulator";
  private static final String ERROR_TAG = "E_IMAGE_MANIPULATOR";
  private ImageLoader mImageLoader;

  public ImageManipulatorModule(Context context) {
    super(context);
  }

  @Override
  public String getName() {
    return TAG;
  }

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mImageLoader = moduleRegistry.getModule(ImageLoader.class);
  }

  @ExpoMethod
  public void manipulateAsync(final String uri, final ArrayList<Object> actions, final ReadableArguments saveOptions, final Promise promise) {
    if (uri == null || uri.length() == 0) {
      promise.reject(ERROR_TAG + "_INVALID_ARG", "Uri passed to ImageManipulator cannot be empty!");
      return;
    }

    final SaveOptions manipulatorSaveOptions;
    final ArrayList<Action> manipulatorActions = new ArrayList<>();
    try {
      manipulatorSaveOptions = SaveOptions.fromArguments(saveOptions);
      for (Object action : actions) {
        manipulatorActions.add(Action.fromObject(action));
      }
    } catch (IllegalArgumentException e) {
      promise.reject(ERROR_TAG + "_INVALID_ARG", e);
      return;
    }

    mImageLoader.loadImageForManipulationFromURL(uri, new ImageLoader.ResultListener() {
          @Override
          public void onSuccess(@NonNull Bitmap bitmap) {
            processBitmapWithActions(bitmap, manipulatorActions, manipulatorSaveOptions, promise);
          }

          @Override
          public void onFailure(@Nullable Throwable cause) {
            // No cleanup required here.
            String basicMessage = "Could not get decoded bitmap of " + uri;
            if (cause != null) {
              promise.reject(ERROR_TAG + "_DECODE",
                  basicMessage + ": " + cause.toString(), cause);
            } else {
              promise.reject(ERROR_TAG + "_DECODE", basicMessage + ".");
            }
          }
        });
  }

  private Bitmap resizeBitmap(Bitmap bitmap, ActionResize resize) {
    float imageRatio = (float) bitmap.getWidth() / bitmap.getHeight();
    int requestedWidth = resize.getWidth() != 0
        ? resize.getWidth()
        : resize.getHeight() != 0
          ? (int) (resize.getHeight() * imageRatio)
          : 0;
    int requestedHeight = resize.getHeight() != 0
        ? resize.getHeight()
        : resize.getWidth() != 0
          ? (int) (resize.getWidth() / imageRatio)
          : 0;
    return Bitmap.createScaledBitmap(bitmap, requestedWidth, requestedHeight, true);
  }

  private Bitmap rotateBitmap(Bitmap bitmap, Integer rotation) {
    Matrix rotationMatrix = new Matrix();
    rotationMatrix.postRotate(rotation);
    return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), rotationMatrix, true);
  }

  private Bitmap flipBitmap(Bitmap bmp, ActionFlip flip) {
    return Bitmap.createBitmap(bmp, 0, 0, bmp.getWidth(), bmp.getHeight(), flip.getRotationMatrix(), true);
  }

  private Bitmap cropBitmap(Bitmap bitmap, ActionCrop crop) throws IllegalArgumentException {
    if (crop.getOriginX() > bitmap.getWidth()
        || crop.getOriginY() > bitmap.getHeight()
        || crop.getOriginX() + crop.getWidth() > bitmap.getWidth()
        || crop.getOriginY() + crop.getHeight() > bitmap.getHeight()
    ) {
      throw new IllegalArgumentException("Invalid crop options has been passed. Please make sure the requested crop rectangle is inside source image.");
    }
    return Bitmap.createBitmap(bitmap, crop.getOriginX(), crop.getOriginY(), crop.getWidth(), crop.getHeight());
  }

  private void processBitmapWithActions(Bitmap bitmap, final ArrayList<Action> actions, final SaveOptions saveOptions, Promise promise) {
    for (Action action : actions) {
      if (action.getResize() != null) {
        bitmap = resizeBitmap(bitmap, action.getResize());
      } else if (action.getRotate() != null) {
        bitmap = rotateBitmap(bitmap, action.getRotate());
      } else if (action.getFlip() != null) {
        bitmap = flipBitmap(bitmap, action.getFlip());
      } else if (action.getCrop() != null) {
        try {
          bitmap = cropBitmap(bitmap, action.getCrop());
        } catch (IllegalArgumentException e) {
          promise.reject(ERROR_TAG + "_CROP_DATA", e);
          return;
        }
      }
    }

    int compression = (int) (saveOptions.getCompress() * 100);

    FileOutputStream out = null;
    ByteArrayOutputStream byteOut = null;
    String path = null;
    String base64String = null;
    try {
      path = FileUtils.generateOutputPath(getContext().getCacheDir(), "ImageManipulator", saveOptions.getFormat().getFileExtension());
      out = new FileOutputStream(path);
      bitmap.compress(saveOptions.getFormat().getCompressFormat(), compression, out);

      if (saveOptions.hasBase64()) {
        byteOut = new ByteArrayOutputStream();
        bitmap.compress(saveOptions.getFormat().getCompressFormat(), compression, byteOut);
        base64String = Base64.encodeToString(byteOut.toByteArray(), Base64.NO_WRAP);
      }
    } catch (Exception e) {
      e.printStackTrace();
    } finally {
      try {
        if (out != null) {
          out.close();
        }
        if (byteOut != null) {
          byteOut.close();
        }
      } catch (IOException e) {
        e.printStackTrace();
      }
    }

    Bundle response = new Bundle();
    response.putString("uri", Uri.fromFile(new File(path)).toString());
    response.putInt("width", bitmap.getWidth());
    response.putInt("height", bitmap.getHeight());
    if (saveOptions.hasBase64()) {
      response.putString("base64", base64String);
    }
    promise.resolve(response);
  }
}

File: packages/expo-image-manipulator/android/src/main/java/expo/modules/imagemanipulator/ImageManipulatorModule.kt
@file:OptIn(EitherType::class)

package expo.modules.imagemanipulator

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.Base64
import expo.modules.imagemanipulator.transformers.CropTransformer
import expo.modules.imagemanipulator.transformers.FlipTransformer
import expo.modules.imagemanipulator.transformers.ResizeTransformer
import expo.modules.imagemanipulator.transformers.RotateTransformer
import expo.modules.interfaces.imageloader.ImageLoaderInterface.ResultListener
import expo.modules.kotlin.apifeatures.EitherType
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.exception.toCodedException
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.sharedobjects.SharedRef
import expo.modules.kotlin.types.EitherOfThree
import expo.modules.kotlin.types.toKClass
import kotlinx.coroutines.async
import kotlinx.coroutines.suspendCancellableCoroutine
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class ImageManipulatorModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private fun createManipulatorContext(url: Uri): ImageManipulatorContext {
    val loader = suspend {
      val imageLoader = appContext.imageLoader
        ?: throw ImageLoaderNotFoundException()

      suspendCancellableCoroutine { continuation ->
        imageLoader.loadImageForManipulationFromURL(
          url.toString(),
          object : ResultListener {
            override fun onSuccess(bitmap: Bitmap) {
              continuation.resume(bitmap)
            }

            override fun onFailure(cause: Throwable?) {
              continuation.resumeWithException(ImageLoadingFailedException(url.toString(), cause.toCodedException()))
            }
          }
        )
      }
    }

    val task = ManipulatorTask(appContext.backgroundCoroutineScope, loader)
    return ImageManipulatorContext(runtimeContext, task)
  }

  private fun createManipulatorContext(bitmap: Bitmap): ImageManipulatorContext {
    val task = ManipulatorTask(appContext.backgroundCoroutineScope) { bitmap }
    return ImageManipulatorContext(runtimeContext, task)
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoImageManipulator")

    Function("manipulate") { url: EitherOfThree<Uri, SharedRef<Bitmap>, SharedRef<Drawable>> ->
      return@Function if (url.`is`(Uri::class)) {
        createManipulatorContext(url.get(Uri::class))
      } else if (url.`is`(toKClass<SharedRef<Bitmap>>())) {
        val bitmap = url.get(toKClass<SharedRef<Bitmap>>()).ref
        createManipulatorContext(bitmap)
      } else {
        val drawable = url.get(toKClass<SharedRef<Drawable>>()).ref
        val bitmap = (drawable as? BitmapDrawable)?.bitmap
          ?: throw Exceptions.IllegalArgument("The drawable cannot be converted to a bitmap")
        createManipulatorContext(bitmap)
      }
    }

    Class<ImageManipulatorContext>("Context") {
      Constructor { url: Uri ->
        createManipulatorContext(url)
      }

      Function("resize") { context: ImageManipulatorContext, options: ResizeOptions ->
        context.addTransformer(ResizeTransformer(options))
      }

      Function("rotate") { context: ImageManipulatorContext, rotation: Float ->
        context.addTransformer(RotateTransformer(rotation))
      }

      Function("flip") { context: ImageManipulatorContext, flipType: FlipType ->
        context.addTransformer(FlipTransformer(flipType))
      }

      Function("crop") { context: ImageManipulatorContext, rect: CropRect ->
        context.addTransformer(CropTransformer(rect))
      }

      Function("reset") { context: ImageManipulatorContext ->
        context.reset()
      }

      AsyncFunction("renderAsync") Coroutine { context: ImageManipulatorContext ->
        val image = context.render()
        ImageRef(image, runtimeContext)
      }
    }

    Class<ImageRef>("Image") {
      Property("width") { image: ImageRef -> image.ref.width }
      Property("height") { image: ImageRef -> image.ref.height }

      AsyncFunction("saveAsync") Coroutine { image: ImageRef, options: ManipulateOptions? ->
        val options = options ?: ManipulateOptions()
        val path = FileUtils.generateRandomOutputPath(context, options.format)
        val compression = (options.compress * 100).toInt()
        val resultBitmap = image.ref

        var base64String: String? = null
        appContext.backgroundCoroutineScope.async {
          FileOutputStream(path).use { fileOut ->
            val compressFormat = options.format.compressFormat
            resultBitmap.compress(compressFormat, compression, fileOut)
            if (options.base64) {
              ByteArrayOutputStream().use { byteOut ->
                resultBitmap.compress(compressFormat, compression, byteOut)
                base64String = Base64.encodeToString(byteOut.toByteArray(), Base64.NO_WRAP)
              }
            }
          }
        }.await()

        mapOf(
          "uri" to Uri.fromFile(File(path)).toString(),
          "width" to resultBitmap.width,
          "height" to resultBitmap.height,
          "base64" to base64String
        )
      }
    }
  }
}
